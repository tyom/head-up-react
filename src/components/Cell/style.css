.cell {
  position: relative;
  opacity: .7;
  transition: .6s;
}

.header {
  padding: 0 .5rem;
  line-height: 1.5rem;
  border-bottom: 1px solid color(#fff a(.1));
  text-transform: uppercase;
  font-size: .8em;
  font-weight: 600;
  position: absolute 0 0 null 0;
  z-index: 1;
}

.menuBtn {
  background: none;
  border: 0;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  color: #fff;
  padding: 0;
  outline: none;
  opacity: 0;
  transition: 200ms opacity;
  margin: 0 5px;

  &:hover .menuIcon {
    opacity: .7;
  }

  &:active .menuIcon {
    opacity: 1;
  }

  > span {
    display: hidden;
  }
}

.menuIcon {
  size: 1.1rem;
  opacity: .4;
}


.container {
  position: absolute 0;
  margin: 4px;
  padding-top: calc(1.5rem + 1px);
  transition: 700ms;
  border: 1px solid color(#fff a(.2));
}

.inner {
  overflow: hidden;
  position: relative;
  height: 100%;
}

.settings,
.content {
  transition: .2s;
  padding: .5rem;
  size: 100%;
}

.content {
}

.settings {
  background-color: #142a59;

  fieldset {
    border: 0;
    padding: 0;
    margin: 0;
  }

  legend {
    margin: 0;
    padding: 0;
    display: block;
    font-weight: bold;
  }
}

* + .field {
  margin-top: 10px;
}

.field label {
  display: table;
}

.field input {
  background: color(#fff a(.2));
  border: 1px solid color(#000 a(.6));
  padding: 2px 4px;
  width: 100%;
  font-size: 1em;

  * + & {
    margin-top: 5px;
  }
}

/* States */

.cell.is-active {
  opacity: 1;

  .container {
    border-color: color(#fff a(.5));
    box-shadow: 0 0 1px 1px color(#fff a(.1));
  }

  .menuBtn {
    opacity: 1;
  }
}

.cell.is-configuring {
  .settings,
  .content {
    transform: translateY(-100%);
  }
}
