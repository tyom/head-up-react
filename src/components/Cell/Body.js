import React from 'react';
import PropTypes from 'prop-types';
import './style.css';

const Body = ({ children, className }) => (
  <div className={className}>{children}</div>
);

Body.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
};

export default Body;
