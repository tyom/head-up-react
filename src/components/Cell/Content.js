import React from 'react';
import PropTypes from 'prop-types';
import './style.css';

const Content = ({ children, className }) => (
  <div styleName="content" className={className}>
    {children}
  </div>
);

Content.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
};

export default Content;
