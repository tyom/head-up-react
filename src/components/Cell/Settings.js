import React from 'react';
import PropTypes from 'prop-types';
import './style.css';

function handleClick(evt, isActive) {
  if (isActive) {
    evt.stopPropagation();
  }
}

const Settings = ({ isActive }) => (
  <aside styleName="settings" onClick={evt => handleClick(evt, isActive)}>
    <fieldset>
      <legend>Settings</legend>
      <div styleName="field">
        <label htmlFor="input">Input</label>
        <input type="text" id="input" />
      </div>
    </fieldset>
  </aside>
);

Settings.propTypes = {
  isActive: PropTypes.bool,
};

export default Settings;
