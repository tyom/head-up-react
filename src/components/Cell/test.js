import React from 'react';
import Cell from './';

describe('Cell: structure', () => {
  let rendered;

  afterEach(() => {
    expect(rendered).toMatchSnapshot();
  });

  test('default component', () => {
    rendered = render(<Cell />);
  });

  test('custom size', () => {
    rendered = render(<Cell size="1:3" />);
  });

  test('inner container class name', () => {
    rendered = render(<Cell innerClassName="useful">content</Cell>).find(
      '.useful'
    );
  });

  test('active state', () => {
    rendered = render(<Cell size="2" isActive={true} />);
  });

  test('activated settings', () => {
    rendered = render(<Cell title="test" isConfiguring={true} />);
  });

  test('with children', () => {
    rendered = render(<Cell size="2">contents</Cell>);
  });
});

describe('Cell: behaviour', () => {
  test('click to activate cell', () => {
    const clickSpy = jest.fn();
    const wrapper = shallow(<Cell onClick={clickSpy} />);

    wrapper.simulate('click');

    expect(clickSpy).toHaveBeenCalled();
  });

  test('shows settings only without children', () => {
    const wrapperWithChildren = mount(
      <Cell title="test" isActive={true}>
        contents
      </Cell>
    );
    const wrapperWithoutChildren = mount(<Cell title="test" isActive={true} />);

    expect(wrapperWithChildren.find('SettingsButton')).toHaveLength(0);
    expect(wrapperWithoutChildren.find('SettingsButton')).toHaveLength(1);
  });

  test('click to activate cell settings', () => {
    const clickSpy = jest.fn();
    const wrapper = mount(
      <Cell title="test" isActive={true} onSettingsClick={clickSpy} />
    );
    const settingsButton = wrapper.find('SettingsButton');

    settingsButton.simulate('click');

    expect(clickSpy).toHaveBeenCalled();
  });
});

describe('Cell: settings', () => {
  test('clicking settings calls a handler', () => {
    const clickSpy = jest.fn();
    const wrapper = shallow(<Cell onClick={clickSpy} />);

    wrapper.simulate('click');

    expect(clickSpy).toHaveBeenCalled();
  });
});
