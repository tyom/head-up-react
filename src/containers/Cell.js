import {connect} from 'react-redux';
import {activateCell} from '../actions';
import {activateCellSettings} from '../actions';

import Cell from '../components/Cell';
import styles from '../components/Cell/styles.css';

const mapStateToProps = (state, ownProps) => ({
  isActive: ownProps.title === state.headup.activeCell,
  isConfiguring: ownProps.title === state.headup.activeCellSettings
});

function mapDispatchToProps(dispatch, ownProps) {
  return {
    onClick: (evt) => {
      const isButton = !!evt.target.closest(`.${styles['Cell-menuBtn']}`);
      if (isButton) {return;}
      dispatch(activateCell(ownProps.title));
    },
    onSettingsClick: (title) => dispatch(activateCellSettings(title))
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Cell);
