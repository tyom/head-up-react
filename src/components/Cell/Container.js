import React from 'react';
import PropTypes from 'prop-types';
import GoSettings from 'react-icons/lib/go/settings';
import Content from './Content';
import Settings from './Settings';

import './style.css';

const SettingsButton = ({ onClick }) => (
  <button styleName="menuBtn" onClick={onClick}>
    <GoSettings styleName="menuIcon" />
    <span>Menu</span>
  </button>
);

SettingsButton.displayName = 'SettingsButton';
SettingsButton.propTypes = {
  onClick: PropTypes.func,
};

const Container = ({
  title,
  isActive,
  onSettingsClick,
  innerClassName,
  children,
}) => (
  <article styleName="container">
    <header styleName="header">
      {title}
      {!children && isActive ? (
        <SettingsButton onClick={() => onSettingsClick(title)} />
      ) : null}
    </header>
    <div styleName="inner">
      <Content className={innerClassName}>{children}</Content>
      {!children ? <Settings isActive={isActive} /> : null}
    </div>
  </article>
);

Container.propTypes = {
  title: PropTypes.string,
  isActive: PropTypes.bool,
  onSettingsClick: PropTypes.func,
  innerClassName: PropTypes.string,
  children: PropTypes.node,
};

export default Container;
