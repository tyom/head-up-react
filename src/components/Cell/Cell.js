import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import Container from './Container';
import { GRID_SIZE } from '../../constants';

import './style.css';

const Cell = ({
  title,
  size = '1',
  isActive,
  isConfiguring,
  onClick,
  onSettingsClick,
  children,
  innerClassName,
  className,
}) => {
  const dimensions = size.split(/:|\//).map(d => `${d / GRID_SIZE * 100}%`);

  return (
    <div
      styleName={classNames('cell', {
        'is-active': isActive,
        'is-configuring': isConfiguring,
      })}
      className={className}
      onClick={onClick}
      style={{
        width: dimensions[0],
        height: dimensions[1] || dimensions[0],
      }}
    >
      {title || children ? (
        <Container
          title={title}
          isActive={isActive}
          onSettingsClick={onSettingsClick}
          innerClassName={innerClassName}
          children={children}
        />
      ) : null}
    </div>
  );
};

Cell.propTypes = {
  onClick: PropTypes.func,
  onSettingsClick: PropTypes.func,
  innerClassName: PropTypes.string,
  size: PropTypes.string,
  title: PropTypes.string,
  isActive: PropTypes.bool,
  isConfiguring: PropTypes.bool,
  children: PropTypes.node,
  className: PropTypes.string,
};

export default Cell;
