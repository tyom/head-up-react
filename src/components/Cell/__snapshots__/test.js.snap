// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell: structure activated settings 1`] = `
<div
  class="style_cell__3Ra6O style_is-configuring__1N8Wb"
  style="width:25%;height:25%"
>
  <article
    class="style_container__1ANRY"
  >
    <header
      class="style_header__3dpY1"
    >
      test
    </header>
    <div
      class="style_inner__1FGTU"
    >
      <div
        class="style_content__1ERGj"
      />
      <aside
        class="style_settings__3Nr_n"
      >
        <fieldset>
          <legend>
            Settings
          </legend>
          <div
            class="style_field__2sqvg"
          >
            <label
              for="input"
            >
              Input
            </label>
            <input
              id="input"
              type="text"
            />
          </div>
        </fieldset>
      </aside>
    </div>
  </article>
</div>
`;

exports[`Cell: structure active state 1`] = `
<div
  class="style_cell__3Ra6O style_is-active__2i7UK"
  style="width:50%;height:50%"
/>
`;

exports[`Cell: structure custom size 1`] = `
<div
  class="style_cell__3Ra6O"
  style="width:25%;height:75%"
/>
`;

exports[`Cell: structure default component 1`] = `
<div
  class="style_cell__3Ra6O"
  style="width:25%;height:25%"
/>
`;

exports[`Cell: structure inner container class name 1`] = `
<div
  class="useful style_content__1ERGj"
>
  content
</div>
`;

exports[`Cell: structure with children 1`] = `
<div
  class="style_cell__3Ra6O"
  style="width:50%;height:50%"
>
  <article
    class="style_container__1ANRY"
  >
    <header
      class="style_header__3dpY1"
    />
    <div
      class="style_inner__1FGTU"
    >
      <div
        class="style_content__1ERGj"
      >
        contents
      </div>
    </div>
  </article>
</div>
`;
